### PR TITLE
Cosmos block manager

### LINKED ISSUE #667: Implement Cosmos block manager
This design is to protect the SwingSet kernel from having to deal with the rewinding, then recalculation of mailbox and kernel state that can happen when called via the Cosmos SDK.

# The Problem

We were naive about what messages should be passed from Cosmos SDK to the SwingSet kernel, relying on the kernel's deduplication of inbound SwingSet messages in order to succeed.

The latest Tendermint (v0.33.0)[https://github.com/tendermint/tendermint/tree/v0.33.0] apparently has the following behaviour:

1. The block proposer creates Block N with its transactions and is asked to produce its state.
2. The Cosmos SDK calls:
  a. BeginBlock (dispatched to modules in SetOrderBeginBlockers)
  b. Transactions in block N (dispatched to individual modules, such as x/swingset)
  c. EndBlock (dispatched to modules in SetOrderEndBlockers)
3. Any keeper changes that were made in the above step are used to calculate Block N's state (app hash).
4. The block proposer signs the resulting state and gossips it to the other validators.
5. Each of the participating validators (**including the block proposer** (new in Tendermint v0.33.0)) are given Block N and are asked to produce the state.
6. The same calls as in steps 2 and 3 are made.
7. The consensus algorithm works to ensure agreement on a given state.  It may repeat from step 1 for the same Block N, but with different transactions.
8. Block N is agreed on and committed, so we repeat from step 1.

Cosmos SDK expects SwingSet to implement a pure function from N + transactions to kernel state.

The main problem vis-a-vis the SwingSet kernel is that the kernel is not a pure function since it incrementally mutates the Javascript heap and we have no efficient way of rewinding state to Block N-1.

# Solution Stages

There are three stages to the solution:

1. Current behaviour: ad-hoc attempts to save state and use kernel message deduplication.  Does not work for the timer device, since it is not idempotent in regards to rerunning the timer with the same timestamp.
2. Accommodate rerunning exactly the same transactions if Block N is repeated (solves only the double-run of the block proposer in steps 2 and 6)
3. Accommodate efficient rewinding of kernel state to Block N-1 (solves the whole issue, but requires snapshots of kernel state)

# Solution Details

Introduce an x/swingset/handler.go handler for EndBlock, which sends an `END_BLOCK` controller call to ag-chain-cosmos.

Introduce an ag-chain-cosmos-specific "Block Manager", which records all SwingSet controller calls (`BEGIN_BLOCK`, `DELIVER_INBOUND`, `END_BLOCK`) from x/swingset/handler.

The Block Manager uses this algorithm (more or less):

```js
let currentMsg;
let currentMessages = [];
let currentHeight;
let replaying = true;
let lastMessages = [];
let lastHeight = 0;
let lastKeeperState;

let newMessages;
let newKeeperState;
function blockManager(msg) {
  if (msg.type === 'BEGIN_BLOCK') {
    // Start a new block.
    replaying = true;
    currentMsg = 0;
    currentMessages = [];
    currentHeight = msg.blockHeight;
  } else {
    // We're working on the next message of the same block.
    currentMsg += 1;
  }

  // Keep track of what we saw.
  currentMessages.push(msg);

  if (!replaying) {
    // Compute new state by running the kernel.
    newKeeperState = kernelPerformMessage(msg);
  } else if (!messagesAreEqual(msg, lastMessages[currentMsg])) {
    // Divergence of the inbound messages, so rewind state if we need to.
    if (currentHeight - 1 !== lastHeight) {
      // For solution stage 2, this call would always throw.
      newKeeperState = setKernelState(currentHeight - 1);
    }
    replaying = false;
    // Resend the messages we have so far to the restored kernel.
    for (const m of currentMessages) {
      newKeeperState = kernelPerformMessage(m);
    }
  }

  if (msg.type === 'END_BLOCK') {
    // Commit all keeper state, even on replay.
    keeperCommit(newKeeperState);

    // Commit the kernel state but only if we didn't replay.
    if (!replaying) {
      kernelCommitState();
    }

    // Advance our stored state variables.
    lastHeight = currentHeight;
    lastMessages = currentMessages;
    lastKeeperState = newKeeperState;
  }
}
```

### PR DESCRIPTION
**Pull request by michaelfig**
Closes #667 

Note that we don't yet have kernel state rewinding.

### CODE REVIEW

**Comment by michaelfig** [comment]
Adding a `COMMIT_BLOCK` event can almost ensure that we don't commit kernel state unless the block has been committed, too.  There is a tiny race here, but only if the block commits and the kernel does not.

Also, panic if the kernel state cannot be rewound. This is crash-only behaviour, which means a rewind attempt will restart the chain from the last committed state (i.e. replay the transcript up to the last committed state).  This should be all we need to clumsily support multiple different proposed blocks at the same height. Eventually, we will hook in with the proper rewinding support to do this more efficiently (and not panic).

File: packages/cosmic-swingset/lib/block-manager.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    return p;
+  }
+
+  let currentActions;
+  let currentIndex;
+  let replaying;
+  let decohered;
+  let saveTime = 0;
+
+  async function blockManager(action) {
```
Comment (warner) [inline packages/cosmic-swingset/lib/block-manager.js:52]: Overall structure looks good, and I know what the `blockManager` is supposed to be doing, but I have to admit I find this control flow hard to follow. We pass multiple times through the same function in different modes, with some code shared between modes and some not.

And it's not clear to me what the starting state is: I'm assuming the first (temporal) thing that happens is a `BEGIN_BLOCK` where both `action.blockHeight` and `savedHeight` are zero, but then the very first thing we do looks like a replay, which seems wrong.

What if, instead of cranking the kernel on each message delivery, we held off cranking it until the `END_BLOCK`? So then all of `BEGIN_BLOCK` and `DELIVER_INBOUND` and `END_BLOCK` would just push actions onto the queue. Then additionally `END_BLOCK` would figure out if we're doing a block for the second time (and compare the contents of the queue against the previously-executed actions), or actually deliver the messages (all at once) and crank the kernel and save the state?

Might that enable a simpler control flow?

**Review by warner — COMMENTED** [review]
probably ok, but if you've got the bandwidth I'd like to see the main `BlockManager` logic be easier to follow

File: packages/cosmic-swingset/lib/block-manager.js (hunk 1 above)
Comment (michaelfig) [reply packages/cosmic-swingset/lib/block-manager.js:52]: I can restructure it a bit better (probably a subfunction for each of the actions) and add comments.  I'm just out of cycles this week.

The sequence is:

```
(BEGIN_BLOCK TX* END_BLOCK)+ COMMIT_BLOCK
```

The only reason I avoided the queue-only, crank at `END_BLOCK` design is that it makes it harder to trace the flow of a Go panic to see what transaction caused a crash.

Comment (michaelfig) [reply packages/cosmic-swingset/lib/block-manager.js:52]: The other reason is that transactions aren't billed to the correct gas meter if we run them all in END_BLOCK.
